src/category_theory/preadditive/projective.lean

```diff
@@ -0,0 +1,202 @@
+/-
+Copyright (c) 2020 Markus Himmel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel, Scott Morrison
+-/
+import algebra.homology.exact
+import category_theory.types
+
+/-!
+# Projective objects and categories with enough projectives
+
+An object `P` is called projective if every morphism out of `P` factors through every epimorphism.
+
+A category `C` has enough projectives if every object admits an epimorphism from some
+projective object.
+
+`projective.over X` picks an arbitrary such projective object,
+and `projective.π X : projective.over X ⟶ X` is the corresponding epimorphism.
+
+Given a morphism `f : X ⟶ Y`, `projective.left f` is a projective object over `kernel f`,
+and `projective.d f : projective.left f ⟶ X` is the morphism `π (kernel f) ≫ kernel.ι f`.
+-/
+
+noncomputable theory
+
+open category_theory
+open category_theory.limits
+
+universes v u
+
+namespace category_theory
+variables {C : Type u} [category.{v} C]
+
+/--
+An object `P` is called projective if every morphism out of `P` factors through every epimorphism.
+-/
+class projective (P : C) : Prop :=
+(factors : ∀ {E X : C} (f : P ⟶ X) (e : E ⟶ X) [epi e], ∃ f', f' ≫ e = f)
+
+section
+
+/--
+A projective presentation of an object `X` consists of an epimorphism `f : P ⟶ X`
+from some projective object `P`.
+-/
+@[nolint has_inhabited_instance]
+structure projective_presentation (X : C) :=
+(P : C)
+(projective : projective P . tactic.apply_instance)
+(f : P ⟶ X)
+(epi : epi f . tactic.apply_instance)
+
+variables (C)
+
+/-- A category "has enough projectives" if for every object `X` there is a projective object `P` and
+    an epimorphism `P ↠ X`. -/
+class enough_projectives : Prop :=
+(presentation : ∀ (X : C), nonempty (projective_presentation X))
+
+end
+
+namespace projective
+
+/--
+An arbitrarily chosen factorisation of a morphism out of a projective object through an epimorphism.
+-/
+def factor_thru {P X E : C} [projective P] (f : P ⟶ X) (e : E ⟶ X) [epi e] : P ⟶ E :=
+(projective.factors f e).some
+
+@[simp] lemma factor_thru_comp {P X E : C} [projective P] (f : P ⟶ X) (e : E ⟶ X) [epi e] :
+  factor_thru f e ≫ e = f :=
+(projective.factors f e).some_spec
+
+section
+local attribute [instance] has_zero_object.has_zero
```

Comment:
Should we change this to `open_locale zero_object` at some point? It's shorter, and (for me) easier to remember.

Reply:
Done.